src/main/java/tfifteenfour/clipboard/model/course/UniqueGroupsList.java

```diff
@@ -0,0 +1,134 @@
+package tfifteenfour.clipboard.model.course;
+
+import static java.util.Objects.requireNonNull;
+import static tfifteenfour.clipboard.commons.util.CollectionUtil.requireAllNonNull;
+
+import java.util.Iterator;
+import java.util.List;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import tfifteenfour.clipboard.model.course.exceptions.DuplicateGroupException;
+import tfifteenfour.clipboard.model.course.exceptions.GroupNotFoundException;
+
+
+public class UniqueGroupsList implements Iterable<Group> {
+
+    private final ObservableList<Group> internalList = FXCollections.observableArrayList();
+    private final ObservableList<Group> internalUnmodifiableList =
+            FXCollections.unmodifiableObservableList(internalList);
+
+    /**
+     * Returns true if the list contains an equivalent student as the given argument.
+     */
+    public boolean contains(Group toCheck) {
+        requireNonNull(toCheck);
+        return internalList.stream().anyMatch(toCheck::isSameGroup);
+    }
+
+    /**
+     * Adds a student to the list.
+     * The student must not already exist in the list.
+     */
+    public void add(Group toAdd) {
+        requireNonNull(toAdd);
+        if (contains(toAdd)) {
+            throw new DuplicateGroupException();
+        }
+        internalList.add(toAdd);
+    }
+
+    /**
+     * Replaces the student {@code target} in the list with {@code editedGroup}.
+     * {@code target} must exist in the list.
+     * The student identity of {@code editedGroup} must not be the same as another existing student in the list.
+     */
+    public void setGroup(Group target, Group editedGroup) {
+        requireAllNonNull(target, editedGroup);
+
+        int index = internalList.indexOf(target);
+        if (index == -1) {
+            throw new GroupNotFoundException();
+        }
+
+        if (!target.isSameGroup(editedGroup) && contains(editedGroup)) {
+            throw new DuplicateGroupException();
+        }
+
+        internalList.set(index, editedGroup);
+    }
+
+    /**
+     * Removes the equivalent student from the list.
+     * The student must exist in the list.
+     */
+    public void remove(Group toRemove) {
+        requireNonNull(toRemove);
+        if (!internalList.remove(toRemove)) {
+            throw new GroupNotFoundException();
+        }
+    }
+
+    public void setGroups(UniqueGroupsList replacement) {
+        requireNonNull(replacement);
+        internalList.setAll(replacement.internalList);
+    }
+
+    /**
+     * Replaces the contents of this list with {@code students}.
+     * {@code students} must not contain duplicate students.
+     */
+    public void setGroups(List<Group> students) {
+        requireAllNonNull(students);
+        if (!personsAreUnique(students)) {
+            throw new DuplicateGroupException();
+        }
+
+        internalList.setAll(students);
+    }
+
```

Comment:
I have similar question on this part also, why would a `List<Group>` contains students

Reply:
Wrong variable name again. Cuz copy pasted the code form UniqueStudentsList >.<